crates/env/src/engine/experimental_off_chain/impls.rs

```diff
@@ -0,0 +1,423 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::EnvInstance;
+use crate::{
+    call::{
+        utils::ReturnType,
+        CallParams,
+        CreateParams,
+    },
+    hash::{
+        Blake2x128,
+        Blake2x256,
+        CryptoHash,
+        HashOutput,
+        Keccak256,
+        Sha2x256,
+    },
+    topics::{
+        Topics,
+        TopicsBuilderBackend,
+    },
+    Clear,
+    EnvBackend,
+    Environment,
+    Result,
+    ReturnFlags,
+    TypedEnvBackend,
+};
+use ink_engine::{
+    ext,
+    ext::Engine,
+};
+use ink_primitives::Key;
+
+const BUFFER_SIZE: usize = 1024;
+
+impl CryptoHash for Blake2x128 {
+    fn hash(input: &[u8], output: &mut <Self as HashOutput>::Type) {
+        type OutputType = [u8; 16];
+        static_assertions::assert_type_eq_all!(
+            <Blake2x128 as HashOutput>::Type,
+            OutputType
+        );
+        let output: &mut OutputType = arrayref::array_mut_ref!(output, 0, 16);
```

Comment:
So I copied this implementation over from our off-chain env and assumed the intention of this part:
```
type OutputType = [u8; 16];
static_assertions::assert_type_eq_all!(
    <Blake2x128 as HashOutput>::Type,
    OutputType
);
let output: &mut OutputType = arrayref::array_mut_ref!(output, 0, 16);
```
was to explicitly ensure that the generic `HashOutput::Type` must always be of `[u8; 16]`. So `OutputType` here will always be equal to `<Self as HashOutput>::Type`, it's just shorter to write.

Reply:
> it's just shorter to write.

and harder to maintain since it does not reuse the implicit constants behind `<<Blake2x128 as HashOutput>::Type as Default>::default()` so I'd ask you to use this more high level API instead.

Reply:
Also we can drop `arrayref` dependency.

Reply:
I'm leaving it as is for now, as we discussed issues with this approach via DM.